apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Destination.java

```diff
@@ -0,0 +1,162 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.impl.context;
+
+import co.elastic.apm.agent.objectpool.Recyclable;
+
+import javax.annotation.Nullable;
+
+/**
+ * Context information about a destination of outgoing calls.
+ */
+public class Destination implements Recyclable {
+
+    /**
+     * An IP (v4 or v6) or a host/domain name.
+     */
+    private final StringBuilder address = new StringBuilder();
+
+    /**
+     * The destination's port used within this context.
+     */
+    private int port;
+
+    public Destination withAddress(@Nullable CharSequence address) {
+        if (address != null && address.length() > 0) {
+            // remove square brackets for IPv6 addresses
+            int startIndex = 0;
+            if (address.charAt(0) == '[') {
+                startIndex = 1;
+            }
+            int endIndex = address.length();
+            if (address.charAt(endIndex - 1) == ']') {
+                endIndex--;
+            }
+            this.address.append(address, startIndex, endIndex);
+        }
+        return this;
+    }
+
+    public StringBuilder getAddress() {
+        return address;
+    }
+
+    public Destination withPort(int port) {
+        this.port = port;
+        return this;
+    }
+
+    public int getPort() {
+        return port;
+    }
+
+    /**
+     * Information about the service related to this destination.
+     */
+    private final Service service = new Service();
+
+    public Service getService() {
+        return service;
+    }
+
+    public boolean hasContent() {
+        return address.length() > 0 || port > 0 || service.hasContent();
+    }
+
+    @Override
+    public void resetState() {
+        address.setLength(0);
+        port = -1;
+        service.resetState();
+    }
+
+    /**
+     * Context information required for service maps.
+     */
+    public static class Service implements Recyclable {
+        /**
+         * Used for detecting unique destinations from each service.
+         * For HTTP, this is the address, with the port (even when it's the default port), without any scheme.
+         * For other types of connections, it's just the {@code span.subtype} (kafka, elasticsearch etc.).
+         * For messaging, we additionally add the queue name (eg jms/myQueue).
+         */
+        private final StringBuilder resource = new StringBuilder();
```

Comment:
What I'd like to see is documentation about which purpose all of these `destination.*` fields exactly serve. Like, where in the UI are they shown and how they are used to produce the service map. Probably the Javadoc is not the best place, but the spec should IMHO explain that.

Reply:
This one is not shown in the UI, it is only used for the query. The `name` is the one shown in the UI, as I wrote in its javadoc. I tried to extract the essence from the GitHub issue and add it here. Once we have a spec, as you suggested, we can refer to it from the javadoc directly.